src/theory/strings/base_solver.cpp

```diff
@@ -0,0 +1,427 @@
+/*********************                                                        */
+/*! \file base_solver.cpp
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Andrew Reynolds
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2019 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Base solver for the theory of strings. This class implements term
+ ** indexing and constant inference for the theory of strings.
+ **/
+
+#include "theory/strings/base_solver.h"
+
+#include "options/strings_options.h"
+#include "theory/strings/theory_strings_rewriter.h"
+#include "theory/strings/theory_strings_utils.h"
+
+using namespace std;
+using namespace CVC4::context;
+using namespace CVC4::kind;
+
+namespace CVC4 {
+namespace theory {
+namespace strings {
+
+BaseSolver::BaseSolver(context::Context* c,
+                       context::UserContext* u,
+                       SolverState& s,
+                       InferenceManager& im)
+    : d_state(s), d_im(im), d_congruent(c)
+{
+  d_emptyString = NodeManager::currentNM()->mkConst(::CVC4::String(""));
+  d_false = NodeManager::currentNM()->mkConst(false);
+}
+
+BaseSolver::~BaseSolver() {}
+
+void BaseSolver::checkInit()
+{
+  // build term index
+  d_eqcToConst.clear();
+  d_eqcToConstBase.clear();
+  d_eqcToConstExp.clear();
+  d_termIndex.clear();
+  d_stringsEqc.clear();
+
+  std::map<Kind, unsigned> ncongruent;
```

Comment:
```suggestion
  std::map<Kind, size_t> ncongruent;
```

Reply:
This is a (debug) count of the number of congruent classes, not an index.  Thus I dont think it should change.

Reply:
Yeah, I was actually hesitating to suggest it in this case. We should probably still make it `uint32_t` to conform [to our guidelines](https://github.com/CVC4/CVC4/wiki/Code-Style-Guidelines#integer-types).